Implement Oneof

wavewave opened the pull request:
Compatible version of protocol-buffers being updated to 2.6.1, this patch enables oneof definition inside message to be defined  as Algebraic Data Type in Haskell. Oneof data type will be generated as a separated module. 
TextMsg and Wire instances are implemented and tested between haskell and C++.

wavewave commented:
This is the result of an effort I have mentioned in #4 .

k-bx commented on hprotoc/Text/ProtocolBuffers/ProtoCompile.hs:
```diff
@@ -252,9 +258,15 @@ run' o@(Output print' writeFile') options env fdps = do
       result = makeResult protoInfo
+  -- trace (concatMap (\x -> show x ++ "\n-------\n") (messages protoInfo)) $ do -- DEBUG
+  -- trace (concatMap (\x -> show x ++ "\n-------\n") (oneofs protoInfo)) $ do -- DEBUG
+  trace (show result) $ do
```
Was tracing left intentionally?

k-bx commented:
@wavewave can you please add a small description into README regarding testing you've added with `build.sh`? Also, I'm getting this error:

```
➜  cpptest git:(wavewave/oneof) ✗ ./build.sh
main.cc:39:1: warning: control reaches end of non-void function [-Wreturn-type]
}
^
main.cc:65:1: warning: control reaches end of non-void function [-Wreturn-type]
}
^
main.cc:88:1: warning: control reaches end of non-void function [-Wreturn-type]
}
^
3 warnings generated.
Undefined symbols for architecture x86_64:
  "std::__1::locale::has_facet(std::__1::locale::id&) const", referenced from:
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::basic_filebuf() in main.o
  "std::__1::locale::use_facet(std::__1::locale::id&) const", referenced from:
      std::__1::basic_ostream<char, std::__1::char_traits<char> >& std::__1::endl<char, std::__1::char_traits<char> >(std::__1::basic_ostream<char, std::__1::char_traits<char> >&) in main.o
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::imbue(std::__1::locale const&) in main.o
      std::__1::basic_ostream<char, std::__1::char_traits<char> >& std::__1::__put_character_sequence<char, std::__1::char_traits<char> >(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, char const*, unsigned long) in main.o
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::basic_filebuf() in main.o
  "std::__1::ios_base::getloc() const", referenced from:
      std::__1::basic_ostream<char, std::__1::char_traits<char> >& std::__1::endl<char, std::__1::char_traits<char> >(std::__1::basic_ostream<char, std::__1::char_traits<char> >&) in main.o
      std::__1::basic_ostream<char, std::__1::char_traits<char> >& std::__1::__put_character_sequence<char, std::__1::char_traits<char> >(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, char const*, unsigned long) in main.o
  "std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::__init(char const*, unsigned long)", referenced from:
      protobuf_AssignDesc_sample_2eproto() in sample.pb.o
  "std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::__init(unsigned long, char)", referenced from:
      std::__1::ostreambuf_iterator<char, std::__1::char_traits<char> > std::__1::__pad_and_output<char, std::__1::char_traits<char> >(std::__1::ostreambuf_iterator<char, std::__1::char_traits<char> >, char const*, char const*, char const*, std::__1::ios_base&, char) in main.o
  "std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::assign(char const*)", referenced from:
      sample_message::set_key(char const*) in main.o
      sample_message::set_name(char const*) in main.o
      sample_message_user::set_name(char const*) in main.o
  "std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::assign(char const*, unsigned long)", referenced from:
      sample_message_user::set_name(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) in sample.pb.o
      sample_message::set_name(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) in sample.pb.o
      sample_message::set_bar(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) in sample.pb.o
      sample_message::set_key(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) in sample.pb.o
  "std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >::~basic_string()", referenced from:
      main1(int, char**) in main.o
      main2(int, char**) in main.o
      main3(int, char**) in main.o
      std::__1::ostreambuf_iterator<char, std::__1::char_traits<char> > std::__1::__pad_and_output<char, std::__1::char_traits<char> >(std::__1::ostreambuf_iterator<char, std::__1::char_traits<char> >, char const*, char const*, char const*, std::__1::ios_base&, char) in main.o
      protobuf_AssignDesc_sample_2eproto() in sample.pb.o
      sample_message_user::SharedDtor() in sample.pb.o
      sample_message::SharedDtor() in sample.pb.o
      ...
  "std::__1::basic_istream<char, std::__1::char_traits<char> >::~basic_istream()", referenced from:
      construction vtable for std::__1::basic_istream<char, std::__1::char_traits<char> >-in-std::__1::basic_ifstream<char, std::__1::char_traits<char> > in main.o
      construction vtable for std::__1::basic_istream<char, std::__1::char_traits<char> >-in-std::__1::basic_fstream<char, std::__1::char_traits<char> > in main.o
  "std::__1::basic_istream<char, std::__1::char_traits<char> >::~basic_istream()", referenced from:
      construction vtable for std::__1::basic_istream<char, std::__1::char_traits<char> >-in-std::__1::basic_ifstream<char, std::__1::char_traits<char> > in main.o
      construction vtable for std::__1::basic_istream<char, std::__1::char_traits<char> >-in-std::__1::basic_fstream<char, std::__1::char_traits<char> > in main.o
  "std::__1::basic_istream<char, std::__1::char_traits<char> >::~basic_istream()", referenced from:
      main2(int, char**) in main.o
      _main in main.o
      std::__1::basic_ifstream<char, std::__1::char_traits<char> >::~basic_ifstream() in main.o
  "std::__1::basic_ostream<char, std::__1::char_traits<char> >::put(char)", referenced from:
      std::__1::basic_ostream<char, std::__1::char_traits<char> >& std::__1::endl<char, std::__1::char_traits<char> >(std::__1::basic_ostream<char, std::__1::char_traits<char> >&) in main.o
  "std::__1::basic_ostream<char, std::__1::char_traits<char> >::flush()", referenced from:
      std::__1::basic_ostream<char, std::__1::char_traits<char> >& std::__1::endl<char, std::__1::char_traits<char> >(std::__1::basic_ostream<char, std::__1::char_traits<char> >&) in main.o
  "std::__1::basic_ostream<char, std::__1::char_traits<char> >::sentry::sentry(std::__1::basic_ostream<char, std::__1::char_traits<char> >&)", referenced from:
      std::__1::basic_ostream<char, std::__1::char_traits<char> >& std::__1::__put_character_sequence<char, std::__1::char_traits<char> >(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, char const*, unsigned long) in main.o
  "std::__1::basic_ostream<char, std::__1::char_traits<char> >::sentry::~sentry()", referenced from:
      std::__1::basic_ostream<char, std::__1::char_traits<char> >& std::__1::__put_character_sequence<char, std::__1::char_traits<char> >(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, char const*, unsigned long) in main.o
  "std::__1::basic_ostream<char, std::__1::char_traits<char> >::~basic_ostream()", referenced from:
      construction vtable for std::__1::basic_ostream<char, std::__1::char_traits<char> >-in-std::__1::basic_fstream<char, std::__1::char_traits<char> > in main.o
  "std::__1::basic_ostream<char, std::__1::char_traits<char> >::~basic_ostream()", referenced from:
      construction vtable for std::__1::basic_ostream<char, std::__1::char_traits<char> >-in-std::__1::basic_fstream<char, std::__1::char_traits<char> > in main.o
  "std::__1::basic_iostream<char, std::__1::char_traits<char> >::~basic_iostream()", referenced from:
      construction vtable for std::__1::basic_iostream<char, std::__1::char_traits<char> >-in-std::__1::basic_fstream<char, std::__1::char_traits<char> > in main.o
  "std::__1::basic_iostream<char, std::__1::char_traits<char> >::~basic_iostream()", referenced from:
      construction vtable for std::__1::basic_iostream<char, std::__1::char_traits<char> >-in-std::__1::basic_fstream<char, std::__1::char_traits<char> > in main.o
  "std::__1::basic_iostream<char, std::__1::char_traits<char> >::~basic_iostream()", referenced from:
      main3(int, char**) in main.o
      _main in main.o
      std::__1::basic_fstream<char, std::__1::char_traits<char> >::~basic_fstream() in main.o
  "std::__1::basic_streambuf<char, std::__1::char_traits<char> >::uflow()", referenced from:
      vtable for std::__1::basic_filebuf<char, std::__1::char_traits<char> > in main.o
  "std::__1::basic_streambuf<char, std::__1::char_traits<char> >::xsgetn(char*, long)", referenced from:
      vtable for std::__1::basic_filebuf<char, std::__1::char_traits<char> > in main.o
  "std::__1::basic_streambuf<char, std::__1::char_traits<char> >::xsputn(char const*, long)", referenced from:
      vtable for std::__1::basic_filebuf<char, std::__1::char_traits<char> > in main.o
  "std::__1::basic_streambuf<char, std::__1::char_traits<char> >::showmanyc()", referenced from:
      vtable for std::__1::basic_filebuf<char, std::__1::char_traits<char> > in main.o
  "std::__1::basic_streambuf<char, std::__1::char_traits<char> >::basic_streambuf()", referenced from:
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::basic_filebuf() in main.o
  "std::__1::basic_streambuf<char, std::__1::char_traits<char> >::~basic_streambuf()", referenced from:
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::~basic_filebuf() in main.o
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::basic_filebuf() in main.o
  "std::__1::cerr", referenced from:
      main2(int, char**) in main.o
      main3(int, char**) in main.o
      _main in main.o
  "std::__1::cout", referenced from:
      main1(int, char**) in main.o
      main2(int, char**) in main.o
      main3(int, char**) in main.o
  "std::__1::ctype<char>::id", referenced from:
      std::__1::basic_ostream<char, std::__1::char_traits<char> >& std::__1::endl<char, std::__1::char_traits<char> >(std::__1::basic_ostream<char, std::__1::char_traits<char> >&) in main.o
      std::__1::basic_ostream<char, std::__1::char_traits<char> >& std::__1::__put_character_sequence<char, std::__1::char_traits<char> >(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, char const*, unsigned long) in main.o
  "std::__1::locale::locale(std::__1::locale const&)", referenced from:
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::basic_filebuf() in main.o
  "std::__1::locale::~locale()", referenced from:
      std::__1::basic_ostream<char, std::__1::char_traits<char> >& std::__1::endl<char, std::__1::char_traits<char> >(std::__1::basic_ostream<char, std::__1::char_traits<char> >&) in main.o
      std::__1::basic_ostream<char, std::__1::char_traits<char> >& std::__1::__put_character_sequence<char, std::__1::char_traits<char> >(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, char const*, unsigned long) in main.o
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::basic_filebuf() in main.o
  "std::__1::codecvt<char, char, __mbstate_t>::id", referenced from:
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::imbue(std::__1::locale const&) in main.o
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::basic_filebuf() in main.o
  "std::__1::ios_base::__set_badbit_and_consider_rethrow()", referenced from:
      std::__1::basic_ostream<char, std::__1::char_traits<char> >& std::__1::__put_character_sequence<char, std::__1::char_traits<char> >(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, char const*, unsigned long) in main.o
  "std::__1::ios_base::init(void*)", referenced from:
      main2(int, char**) in main.o
      main3(int, char**) in main.o
      _main in main.o
  "std::__1::ios_base::clear(unsigned int)", referenced from:
      main3(int, char**) in main.o
      _main in main.o
      std::__1::basic_ifstream<char, std::__1::char_traits<char> >::open(char const*, unsigned int) in main.o
      std::__1::basic_ostream<char, std::__1::char_traits<char> >& std::__1::__put_character_sequence<char, std::__1::char_traits<char> >(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, char const*, unsigned long) in main.o
  "std::__1::basic_ios<char, std::__1::char_traits<char> >::~basic_ios()", referenced from:
      main2(int, char**) in main.o
      main3(int, char**) in main.o
      _main in main.o
      std::__1::basic_ifstream<char, std::__1::char_traits<char> >::~basic_ifstream() in main.o
      std::__1::basic_fstream<char, std::__1::char_traits<char> >::~basic_fstream() in main.o
  "std::bad_cast::bad_cast()", referenced from:
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::seekoff(long long, std::__1::ios_base::seekdir, unsigned int) in main.o
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::sync() in main.o
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::underflow() in main.o
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::overflow(int) in main.o
  "std::bad_cast::~bad_cast()", referenced from:
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::seekoff(long long, std::__1::ios_base::seekdir, unsigned int) in main.o
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::sync() in main.o
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::underflow() in main.o
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::overflow(int) in main.o
  "std::terminate()", referenced from:
      ___clang_call_terminate in main.o
      ___clang_call_terminate in sample.pb.o
  "typeinfo for std::__1::basic_istream<char, std::__1::char_traits<char> >", referenced from:
      construction vtable for std::__1::basic_istream<char, std::__1::char_traits<char> >-in-std::__1::basic_ifstream<char, std::__1::char_traits<char> > in main.o
      typeinfo for std::__1::basic_ifstream<char, std::__1::char_traits<char> > in main.o
      construction vtable for std::__1::basic_istream<char, std::__1::char_traits<char> >-in-std::__1::basic_fstream<char, std::__1::char_traits<char> > in main.o
  "typeinfo for std::__1::basic_ostream<char, std::__1::char_traits<char> >", referenced from:
      construction vtable for std::__1::basic_ostream<char, std::__1::char_traits<char> >-in-std::__1::basic_fstream<char, std::__1::char_traits<char> > in main.o
  "typeinfo for std::__1::basic_iostream<char, std::__1::char_traits<char> >", referenced from:                                                                                                                                                                                                                                                                     [12/902]
      construction vtable for std::__1::basic_iostream<char, std::__1::char_traits<char> >-in-std::__1::basic_fstream<char, std::__1::char_traits<char> > in main.o
      typeinfo for std::__1::basic_fstream<char, std::__1::char_traits<char> > in main.o
  "typeinfo for std::__1::basic_streambuf<char, std::__1::char_traits<char> >", referenced from:
      typeinfo for std::__1::basic_filebuf<char, std::__1::char_traits<char> > in main.o
  "typeinfo for std::bad_cast", referenced from:
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::seekoff(long long, std::__1::ios_base::seekdir, unsigned int) in main.o
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::sync() in main.o
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::underflow() in main.o
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::overflow(int) in main.o
  "vtable for __cxxabiv1::__si_class_type_info", referenced from:
      typeinfo for std::__1::basic_ifstream<char, std::__1::char_traits<char> > in main.o
      typeinfo for std::__1::basic_filebuf<char, std::__1::char_traits<char> > in main.o
      typeinfo for std::__1::basic_fstream<char, std::__1::char_traits<char> > in main.o
      typeinfo for sample_message_user in sample.pb.o
      typeinfo for sample_message_simple in sample.pb.o
      typeinfo for sample_message in sample.pb.o
  NOTE: a missing vtable usually means the first non-inline virtual member function has no definition.
  "vtable for std::__1::ios_base", referenced from:
      main2(int, char**) in main.o
      main3(int, char**) in main.o
      _main in main.o
  NOTE: a missing vtable usually means the first non-inline virtual member function has no definition.
  "vtable for std::__1::basic_ios<char, std::__1::char_traits<char> >", referenced from:
      main2(int, char**) in main.o
      main3(int, char**) in main.o
      _main in main.o
  NOTE: a missing vtable usually means the first non-inline virtual member function has no definition.
  "non-virtual thunk to std::__1::basic_iostream<char, std::__1::char_traits<char> >::~basic_iostream()", referenced from:
      construction vtable for std::__1::basic_iostream<char, std::__1::char_traits<char> >-in-std::__1::basic_fstream<char, std::__1::char_traits<char> > in main.o
  "non-virtual thunk to std::__1::basic_iostream<char, std::__1::char_traits<char> >::~basic_iostream()", referenced from:
      construction vtable for std::__1::basic_iostream<char, std::__1::char_traits<char> >-in-std::__1::basic_fstream<char, std::__1::char_traits<char> > in main.o
  "non-virtual thunk to std::__1::basic_istream<char, std::__1::char_traits<char> >::~basic_istream()", referenced from:
      construction vtable for std::__1::basic_istream<char, std::__1::char_traits<char> >-in-std::__1::basic_ifstream<char, std::__1::char_traits<char> > in main.o
      construction vtable for std::__1::basic_istream<char, std::__1::char_traits<char> >-in-std::__1::basic_fstream<char, std::__1::char_traits<char> > in main.o
  "non-virtual thunk to std::__1::basic_istream<char, std::__1::char_traits<char> >::~basic_istream()", referenced from:
      construction vtable for std::__1::basic_istream<char, std::__1::char_traits<char> >-in-std::__1::basic_ifstream<char, std::__1::char_traits<char> > in main.o
      construction vtable for std::__1::basic_istream<char, std::__1::char_traits<char> >-in-std::__1::basic_fstream<char, std::__1::char_traits<char> > in main.o
  "non-virtual thunk to std::__1::basic_ostream<char, std::__1::char_traits<char> >::~basic_ostream()", referenced from:
      construction vtable for std::__1::basic_ostream<char, std::__1::char_traits<char> >-in-std::__1::basic_fstream<char, std::__1::char_traits<char> > in main.o
  "non-virtual thunk to std::__1::basic_ostream<char, std::__1::char_traits<char> >::~basic_ostream()", referenced from:
      construction vtable for std::__1::basic_ostream<char, std::__1::char_traits<char> >-in-std::__1::basic_fstream<char, std::__1::char_traits<char> > in main.o
  "non-virtual thunk to std::__1::basic_iostream<char, std::__1::char_traits<char> >::~basic_iostream()", referenced from:
      construction vtable for std::__1::basic_iostream<char, std::__1::char_traits<char> >-in-std::__1::basic_fstream<char, std::__1::char_traits<char> > in main.o
  "non-virtual thunk to std::__1::basic_iostream<char, std::__1::char_traits<char> >::~basic_iostream()", referenced from:
      construction vtable for std::__1::basic_iostream<char, std::__1::char_traits<char> >-in-std::__1::basic_fstream<char, std::__1::char_traits<char> > in main.o
  "operator delete[](void*)", referenced from:
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::~basic_filebuf() in main.o
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::imbue(std::__1::locale const&) in main.o
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::setbuf(char*, long) in main.o
      void google::protobuf::internal::RepeatedPtrFieldBase::Destroy<google::protobuf::RepeatedPtrField<sample_message_user>::TypeHandler>() in sample.pb.o
  "operator delete(void*)", referenced from:
      main2(int, char**) in main.o
      _main in main.o
      google::protobuf::internal::GenericTypeHandler<sample_message_user>::New() in main.o
      std::__1::basic_ifstream<char, std::__1::char_traits<char> >::~basic_ifstream() in main.o
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::~basic_filebuf() in main.o
      std::__1::basic_fstream<char, std::__1::char_traits<char> >::~basic_fstream() in main.o
      protobuf_AssignDesc_sample_2eproto() in sample.pb.o
      ...
  "operator new[](unsigned long)", referenced from:
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::imbue(std::__1::locale const&) in main.o
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::setbuf(char*, long) in main.o
  "operator new(unsigned long)", referenced from:
      main2(int, char**) in main.o
      _main in main.o
      sample_message::set_key(char const*) in main.o
      sample_message::set_name(char const*) in main.o
      sample_message_user::set_name(char const*) in main.o
      google::protobuf::internal::GenericTypeHandler<sample_message_user>::New() in main.o
      protobuf_AssignDesc_sample_2eproto() in sample.pb.o
      ...
  "___cxa_allocate_exception", referenced from:
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::seekoff(long long, std::__1::ios_base::seekdir, unsigned int) in main.o
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::sync() in main.o
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::underflow() in main.o
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::overflow(int) in main.o
  "___cxa_begin_catch", referenced from:
      ___clang_call_terminate in main.o
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::~basic_filebuf() in main.o
      std::__1::basic_ostream<char, std::__1::char_traits<char> >& std::__1::__put_character_sequence<char, std::__1::char_traits<char> >(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, char const*, unsigned long) in main.o
      ___clang_call_terminate in sample.pb.o
  "___cxa_call_unexpected", referenced from:
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::close() in main.o
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::basic_filebuf() in main.o
  "___cxa_end_catch", referenced from:
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::~basic_filebuf() in main.o
      std::__1::basic_ostream<char, std::__1::char_traits<char> >& std::__1::__put_character_sequence<char, std::__1::char_traits<char> >(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, char const*, unsigned long) in main.o
  "___cxa_guard_acquire", referenced from:
      protobuf_AssignDesc_sample_2eproto() in sample.pb.o
  "___cxa_guard_release", referenced from:
      protobuf_AssignDesc_sample_2eproto() in sample.pb.o
  "___cxa_throw", referenced from:
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::seekoff(long long, std::__1::ios_base::seekdir, unsigned int) in main.o
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::sync() in main.o
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::underflow() in main.o
      std::__1::basic_filebuf<char, std::__1::char_traits<char> >::overflow(int) in main.o
  "___dynamic_cast", referenced from:
      sample_message_user const* google::protobuf::internal::dynamic_cast_if_available<sample_message_user const*, google::protobuf::Message const*>(google::protobuf::Message const*) in sample.pb.o
      sample_message_simple const* google::protobuf::internal::dynamic_cast_if_available<sample_message_simple const*, google::protobuf::Message const*>(google::protobuf::Message const*) in sample.pb.o
      sample_message const* google::protobuf::internal::dynamic_cast_if_available<sample_message const*, google::protobuf::Message const*>(google::protobuf::Message const*) in sample.pb.o
  "___gxx_personality_v0", referenced from:
      main1(int, char**) in main.o
      main2(int, char**) in main.o
      main3(int, char**) in main.o
      _main in main.o
      std::__1::basic_ostream<char, std::__1::char_traits<char> >& std::__1::endl<char, std::__1::char_traits<char> >(std::__1::basic_ostream<char, std::__1::char_traits<char> >&) in main.o
      std::__1::basic_ifstream<char, std::__1::char_traits<char> >::~basic_ifstream() in main.o
      std::__1::basic_fstream<char, std::__1::char_traits<char> >::~basic_fstream() in main.o
      ...
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

Can you help me out please?

wavewave commented:
I removed unnecessary debug traces. 

For cpptest code compilation, I added pkg-config --cflags protobuf for compilation.
The error looks related to libc++ or libstdc++ difference between clang and g++. If the problem persists, try to add -stdlib=libc++ or -stdlib=libstdc++. I will also find some machine with clang and test it.

wavewave commented:
@k-bx I will add testing to README. In the long run, it is good to have standard to do inter-language tests.

k-bx commented:
@wavewave it errors for me on ubuntu:

```
kb@mini:~/workspace/protocol-buffers/hprotoc/cpptest$ ./build.sh
/usr/bin/ld: sample.pb.o: undefined reference to symbol '_ZdaPv@@GLIBCXX_3.4'
//usr/lib/x86_64-linux-gnu/libstdc++.so.6: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
```

Replacing "gcc" for "g++" works. Can you do the replacement please?

k-bx commented:
@wavewave also, what should one do to test after "build.sh" succeeds?

wavewave commented:
I've clean up the test code implementation and explain this in README.md (See the last part) 
The test code is located in `hprotoc/oneoftest` and under that, we have `school.proto` and `hs` directory for haskell, and `cpp` directory for C++ decode/encode example. 
The `school.proto` example describes a school with dormitory system that can contain members as one of faculty, students and admin. I just use some fun Harry Potter example.
Let me know if anything is not clear. Thanks!

k-bx commented:
@wavewave this is amazing! Thank you very much!
